### ISSUE TITLE
[Bug] the paypal Frame flicker when use onClick to validate form

### ISSUE THREAD

**Issue by 2632755669** [paypal/paypal-js#449]
### 🐞 Describe the Bug

`           onClick: function (data, actions) {
                if (!validateForm()) {
                  return actions.reject();
                };
            },
`
Its purpose is to prevent the user from making the next payment if the verification form fails， But this will cause a Frame flicker problem；

### 😕 Actual Behavior

1、       onClick: function (data, actions) {
              if (!validateForm()) {
                return actions.reject();
              };
            },

### 🤔 Expected Behavior
it shouldn't show the Frame and close when i click paypal button and execute actions.reject() in the onClick

### 🌍 Environment

-   Node.js/npm: -
-   OS: -
-   Browser: - chrome

**Comment by jshawl** [comment]
Another way to handle form validation would be to use `actions.disable()` inside `onInit` and listen for events that validate the form, and once valid, call `actions.enable()`

Here is an example:

```js
paypal.Buttons({
  onInit(data, actions) {
    actions.disable(); // disabled, by default
    document.querySelector("form").addEventListener("change", (e) => {
      if (validateForm()) {
        actions.enable();
      } else {
        actions.disable();
      }
    });
  },
  onClick(data, actions) {
    if (!validateForm()) {
      // show an error message on the form
    }
    // no other code is needed here for the click to initiate the popup checkout flow
  }
}).render("#paypal-button-container");
```

This page includes complete examples for both client-side and server-side form validation:  https://developer.paypal.com/docs/checkout/standard/customize/validate-user-input/

`actions.reject()` is best used for preventing a payment based on what the buyer enters within the popup, before the payment is approved, e.g. changing their shipping address to an unsupported country. (https://developer.paypal.com/sdk/js/reference/#link-examples)

Let us know if this doesn't solve the issue and we can take another look, thanks!

**Comment by knoefel** [comment]
@jshawl @wsbrunson  I think a common use case is (at least its our use case 😁), that you want to trigger the validation after the user has clicked the paypal submit button. And only after the validation succeeded the popup should open. We are using the adyen integration for paypal and experience a similar flicker of the popup. Any ideas what we can do in this scenario? Here is our react code for setting up the paypal button:

```ts
paypal
  .create('paypal', {
    style: { color: 'gold', shape: 'pill', label: 'pay' },
    onClick: async () => {
      const { form, paypal } = get()

      if (!form?.current) return

      form.current.setSubmitting(true)

      const errors = await form.current.validateForm()
      if (!isEmpty(errors)) {
        await form.current.setTouched(setNestedObjectValues(errors, true))

        return false
      }

      paypal?.setState({ input: getInput(form.current.values) })
    },
  })
  .mount(nodes.paypal)
```

Thanks in advance!

**Comment by wsbrunson** [comment]
@knoefel I agree completely, this is a common pain-point that comes up with paypal integrations.

Unfortunately, this is not a limitation of the PayPal SDK but the browsers themselves. Once the browsers detect a click, we have less than 1 second (the rules in the browsers can be vague but it basically needs to be immediately) to open the popup window or else it will trigger the browser's adblock and prevent it from opening.

This is why that flicker happens, we have to open it immediately. This is also why the popup first opens to `about:blank`. We immediately open the popup, run `createOrder`, then redirect to our checkout app with the order id once `createOrder` resolves.

**Comment by knoefel** [comment]
@wsbrunson Thanks a lot for the quick and useful reply!!

**Comment by Basheer-alsyani** [comment]
Basheer Ahmed ali alsyani

**Comment by Basheer-alsyani** [comment]
VR8649ZEEG4UL

**Comment by Jelly38214** [comment]
Hi guys, is there any update for solving the scenario that @knoefel mentioned?

**Comment by wsbrunson** [comment]
hi @Jelly38214,

We don't have a way to solve the popup flash when validating onClick because of the nature of 3rd-party popups and User Activation restrictions in browsers.

We currently do not provide a javascript-only mechanism for triggering the PayPal button. In the future, we plan to offer a more flexible integration path for developers, including providing a way to open the Checkout window through javascript. It is unlikely that these changes will happen in the current generation of our SDK though.

I can't speak to any timelines on when this next generation of our SDK will be available, I would make sure to check our developer announcements throughout this year

**Comment by boy51** [comment]
Wait how is this a browser limitation? What we are trying to achieve is very simple. NOT open a popup after onClick if we want to. How are you saying this is only possible with an all new PayPal sdk? There is no issue with user activation because nothing is supposed to happen.

**Comment by wsbrunson** [comment]
> Wait how is this a browser limitation?

Once the button is clicked, we have to open the popup immediately after. We can't wait for even a next tick because the PayPal SDK is third-party javascript running on a merchant's page. If we didn't open immediately, the browser would consider this a popup ad or some type of unwanted popup and block the window from opening.

In the negative case, yes this would work as you described. You make a check in `onClick` and decide not to continue and the popup never opens

In the positive case though, it may not work. You make a check in `onClick` and decide to continue with the transaction but your check took too long. Now, when the PayPal SDK goes to open the new window, the browser determines that too much time has elapsed between the button click and the opening of the window and blocks opening that window.

Once we provide a javascript-only mechanism for opening the checkout window, the code that the merchant owns will decide what to do on click and whether to open the popup or not. The merchant's own code running on their site is not subject to the same restrictions as the PayPal SDK is.
